multishard_mutation_query.cc

```diff
@@ -456,34 +457,29 @@ read_context::dismantle_buffer_stats read_context::dismantle_compaction_state(de
         return stats;
     }
 
-    auto& shard_buffer = _readers[shard].get_dismantled_buffer(_permit);
+    _readers[shard].compaction_state = std::move(compaction_state);
 
     if (rtc_opt) {
         stats.add(*_schema, *rtc_opt);
-        shard_buffer.emplace_front(*_schema, _permit, std::move(*rtc_opt));
     }
-
     if (compaction_state.static_row) {
         stats.add(*_schema, *compaction_state.static_row);
-        shard_buffer.emplace_front(*_schema, _permit, std::move(*compaction_state.static_row));
     }
-
     stats.add(*_schema, compaction_state.partition_start);
-    shard_buffer.emplace_front(*_schema, _permit, std::move(compaction_state.partition_start));
 
     return stats;
 }
 
 future<> read_context::save_reader(shard_id shard, const dht::decorated_key& last_pkey, position_in_partition_view last_pos) {
   return do_with(std::exchange(_readers[shard], {}), [this, shard, &last_pkey, last_pos] (reader_meta& rm) mutable {
     return _db.invoke_on(shard, [this, query_uuid = _cmd.query_uuid, query_ranges = _ranges, &rm,
-            &last_pkey, last_pos, gts = tracing::global_trace_state_ptr(_trace_state)] (replica::database& db) mutable {
+            &last_pkey, last_pos, gts = tracing::global_trace_state_ptr(_trace_state)] (replica::database& db) mutable -> future<> {
         try {
             auto rparts = rm.rparts.release(); // avoid another round-trip when destroying rparts
             auto reader_opt = rparts->permit.semaphore().unregister_inactive_read(std::move(*rparts->handle));
 
             if (!reader_opt) {
-                return make_ready_future<>();
+                co_return;
```

Comment:
Please make this a non-coroutine again, I expect we'll need a backport, and as far as I can tell the change isn't required.

Reply:
Well, there's a co_await later on. hmm.

Reply:
We can switch to continuations when backporting, so master will keep the pretty code, only the older versions will suffer the  continuation-added-in-the-middle.

Reply:
Yes. But I think switch to looking at the last fragment is also simpler.